service/controller/app/v1/resource/tcnamespace/create.go

```diff
@@ -53,10 +55,40 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating namespace %#q in tenant cluster %#q", ns.Name, key.ClusterID(cr)))
 
-	_, err = cc.K8sClient.CoreV1().Namespaces().Create(ns)
+	ch := make(chan error)
+
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	go func() {
+		_, err = cc.K8sClient.CoreV1().Namespaces().Create(ns)
+		ch <- err
+	}()
+
+	select {
+	case err = <-ch:
+		// Fall through.
+	case <-ctx.Done():
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			// Set status so we don't try to connect to the tenant cluster
+			// again in this reconciliation loop.
+			cc.Status.TenantCluster.IsUnavailable = true
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "timeout creating namespace")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		}
+	default:
```

Comment:
I don't think we need `default` in here since in any case `ch` is not yet ready, then it would fall into here.

Reply:
Thanks! Fixed.